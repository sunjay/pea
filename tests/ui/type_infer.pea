fn main() {
    // ambiguous type
    let a = [];

    // non-homogenous list
    let y = [3, 4, b"abc"];
    // does not match type annotation on variable
    let z: [i64] = [b"a", b"c"];

    // does not match function signature
    let q = [b""];
    let r: [u8] = foo(q);

    let f = foo;
    // does not match the signature of the first function assigned to f
    f = bar;

    // non-boolean if condition
    if foo([1]) {
        // println can contain anything
        println!(33);
        // println can contain anything
        println!(b"abc");

        // Return expression is fine in a block as long as it evaluates to () or !
        ()
    } else {
        // Return expression that evaluates to !
        return ()
    }

    // condition of while loop must be a boolean
    while b"test test test" {
        println!(f);

        // body of while loop must evaluate to `()`
        1
    }

    loop {
        println!(3);

        if false {
            // `continue` can be any type since it never actually returns
            let v: i64 = continue;
        } else {
            // `break` can be any type since it never actually returns
            let v: [u8] = break;
        }

        // Both `continue` and `break` should continue to type check as `()`, even if there is
        // nothing to disambiguate their types
        if false {
            continue;
            // implicit `()`
        } else {
            break;
            // implicit `()`
        }

        // body of loop must evaluate to `()`
        b"ok"
    }

    // all branches of if must produce the same type
    let var1 = if false {
        55
    } else if true {
        b""
    } else {
        -333
    };

    // if without an else must produce `()`
    let var2 = if true {
        45
    };

    // Complex expressions with errors
    let x = (4 * 3) - 45 / b"";
    let y = b"" % bar(3333);

    // Non-function call
    bar(-111)(4);

    let x = if false {
        // return evaluates to any type but return expression type must be the return type of the
        // function
        return 33
    } else {
        44
    };

    // return with no argument returns `()`
    return
}

fn foo(x: [i64]) -> i64 {
    3 * 4
}

fn bar(x: i64) -> [u8] {
    b"test"
}

fn returns_unit() {
    if false {
        // Not a value of type `()`
        return b"";
    }

    // Not a value of type `()`
    3
}
