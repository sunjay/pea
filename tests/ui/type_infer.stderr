[tests/ui/type_infer.pea:6:17] error: mismatched types: expected `i64`, found: `[u8]`

[tests/ui/type_infer.pea:8:20-8:31] error: mismatched types: expected `i64`, found: `[u8]`

[tests/ui/type_infer.pea:8:20-8:31] error: mismatched types: expected `i64`, found: `[u8]`

[tests/ui/type_infer.pea:11:14-11:16] error: mismatched types: expected `[u8]`, found: `i64`

[tests/ui/type_infer.pea:98:4-98:6] error: mismatched types: expected `[i64]`, found: `i64`

[tests/ui/type_infer.pea:19:8-19:10] error: mismatched types: expected `bool`, found: `i64`

[tests/ui/type_infer.pea:33:11-33:27] error: mismatched types: expected `bool`, found: `[u8]`

[tests/ui/type_infer.pea:37:9] error: mismatched types: expected `()`, found: `i64`

[tests/ui/type_infer.pea:62:9-62:13] error: mismatched types: expected `()`, found: `[u8]`

[tests/ui/type_infer.pea:67:9-67:10] error: mismatched types: expected `i64`, found: `[u8]`

[tests/ui/type_infer.pea:76:9-76:10] error: mismatched types: expected `i64`, found: `()`

[tests/ui/type_infer.pea:80:23-80:24] error: mismatched types: expected `i64`, found: `[u8]`

[tests/ui/type_infer.pea:84:5-84:13] error: mismatched types: expected `fn (_)-> _`, found: `[u8]`

[tests/ui/type_infer.pea:29:16-29:17] error: mismatched types: expected `()`, found: `i64`

[tests/ui/type_infer.pea:109:16-109:18] error: mismatched types: expected `()`, found: `[u8]`

[tests/ui/type_infer.pea:113:5] error: mismatched types: expected `()`, found: `i64`

error: aborting due to 16 previous errors

